### ISSUE TITLE
Regarding the model reproduction issue

### ISSUE THREAD

**Issue by typist2001** [hanjun815/SHeRLoc#2]
Regarding the model reproduction issue:
We have attempted to train your model, and below are the results of the training process:
root@server:/code/SHeRLoc/training# python train.py --config ../config/config.txt --model_config ../models/Hetero_SHeRLoc.txt
Training config path: ../config/config.txt
Model config path: ../models/Hetero_SHeRLoc.txt
Debug mode: False
Parameters:
params_path: ../config/config.txt
model_params_path: ../models/Hetero_SHeRLoc.txt
dataset: hercules
dataset_folder: /code/SHeRLoc/datasets/HeRCULES
num_workers: 1
batch_size: 20
lr: 0.001
epochs: 20
weight_decay: 0.001
loss: BatchHardTripletMarginLoss
margin: 1.0
train_dataset: /code/SHeRLoc/datasets/HeRCULES
val_dataset: /code/SHeRLoc/datasets/HeRCULES
test_dataset: /code/SHeRLoc/datasets/HeRCULES
Model parameters:
model_params_path: ../models/Hetero_SHeRLoc.txt
task: Hetero
radar_model: SHeRLoc

Model name: model_SHeRLoc_20251104_1435
Number of model parameters: 19529301
11710 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2342 readings from sensor: png with valid pose, 0 rejected due to unknown pose
9319 readings from sensor: png with valid pose, 0 rejected due to unknown pose
1864 readings from sensor: png with valid pose, 0 rejected due to unknown pose
7805 readings from sensor: png with valid pose, 0 rejected due to unknown pose
1561 readings from sensor: png with valid pose, 0 rejected due to unknown pose
11630 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2331 readings from sensor: png with valid pose, 0 rejected due to unknown pose
19575 readings from sensor: png with valid pose, 0 rejected due to unknown pose
3930 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2510 readings from sensor: png with valid pose, 0 rejected due to unknown pose
502 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2237 readings from sensor: png with valid pose, 0 rejected due to unknown pose
447 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2321 readings from sensor: png with valid pose, 0 rejected due to unknown pose
464 readings from sensor: png with valid pose, 0 rejected due to unknown pose
2198 readings from sensor: png with valid pose, 0 rejected due to unknown pose
440 readings from sensor: png with valid pose, 0 rejected due to unknown pose
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Tracking run with wandb version 0.22.3
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
wandb: Run data is saved locally in /code/SHeRLoc/training/wandb/offline-run-20251104_143519-d31bl2yx
  0%|                                                                                                                                                                                               | 0/20 [00:00<?, ?it/s]train - Global loss: 0.134771    Embedding norm: 1.0000   Triplets (all/active): 20.0/139.8
Pos dist (min/mean/max): 0.7435/0.9687/1.2117   Neg dist (min/mean/max): 1.1242/1.3944/1.6378
val - Global loss: 0.357243    Embedding norm: 1.0000   Triplets (all/active): 20.0/337.3
Pos dist (min/mean/max): 0.5188/0.8079/1.2276   Neg dist (min/mean/max): 0.4819/0.9330/1.7741
  5%|█████████                                                                                                                                                                           | 1/20 [13:06<4:09:08, 786.75s/it]train - Global loss: 0.046871    Embedding norm: 1.0000   Triplets (all/active): 20.0/36.9
Pos dist (min/mean/max): 0.5053/0.6482/0.8272   Neg dist (min/mean/max): 1.0646/1.2104/1.4045
val - Global loss: 0.305816    Embedding norm: 1.0000   Triplets (all/active): 20.0/368.0
Pos dist (min/mean/max): 0.4443/0.6592/0.8905   Neg dist (min/mean/max): 0.4104/0.7488/1.5411
 10%|██████████████████                                                                                                                                                                  | 2/20 [25:16<3:45:54, 753.02s/it]train - Global loss: 0.053715    Embedding norm: 1.0000   Triplets (all/active): 20.0/47.6
Pos dist (min/mean/max): 0.3661/0.4709/0.6181   Neg dist (min/mean/max): 0.8650/0.9934/1.1366
val - Global loss: 0.224425    Embedding norm: 1.0000   Triplets (all/active): 20.0/372.3
Pos dist (min/mean/max): 0.3248/0.4695/0.6698   Neg dist (min/mean/max): 0.3139/0.6003/1.4279
 15%|███████████████████████████                                                                                                                                                         | 3/20 [37:24<3:30:11, 741.85s/it]train - Global loss: 0.017706    Embedding norm: 1.0000   Triplets (all/active): 20.0/13.2
Pos dist (min/mean/max): 0.2529/0.3341/0.4556   Neg dist (min/mean/max): 0.8389/0.9362/1.0647
val - Global loss: 0.220599    Embedding norm: 1.0000   Triplets (all/active): 20.0/376.5
Pos dist (min/mean/max): 0.2900/0.4288/0.6041   Neg dist (min/mean/max): 0.3053/0.5598/1.3213
 20%|████████████████████████████████████                                                                                                                                                | 4/20 [49:32<3:16:23, 736.47s/it]train - Global loss: 0.036106    Embedding norm: 1.0000   Triplets (all/active): 20.0/45.0
Pos dist (min/mean/max): 0.1876/0.2659/0.3843   Neg dist (min/mean/max): 0.7161/0.8558/0.9878
val - Global loss: 0.146573    Embedding norm: 1.0000   Triplets (all/active): 20.0/395.4
Pos dist (min/mean/max): 0.1300/0.1715/0.2240   Neg dist (min/mean/max): 0.1210/0.2087/0.3946
 25%|████████████████████████████████████████████▌                                                                                                                                     | 5/20 [1:01:39<3:03:15, 733.05s/it]train - Global loss: 0.036728    Embedding norm: 1.0000   Triplets (all/active): 20.0/38.3
Pos dist (min/mean/max): 0.0938/0.1644/0.2672   Neg dist (min/mean/max): 0.5814/0.7898/0.9547
val - Global loss: 0.161366    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.4
Pos dist (min/mean/max): 0.0244/0.0705/0.1625   Neg dist (min/mean/max): 0.0237/0.0766/0.2102
 30%|█████████████████████████████████████████████████████▍                                                                                                                            | 6/20 [1:13:41<2:50:10, 729.29s/it]train - Global loss: 0.015326    Embedding norm: 1.0000   Triplets (all/active): 20.0/14.3
Pos dist (min/mean/max): 0.0569/0.1075/0.1886   Neg dist (min/mean/max): 0.6895/0.8269/0.9125
val - Global loss: 0.215614    Embedding norm: 1.0000   Triplets (all/active): 20.0/399.2
Pos dist (min/mean/max): 0.0200/0.1135/0.2323   Neg dist (min/mean/max): 0.0195/0.1206/0.6676
 35%|██████████████████████████████████████████████████████████████▎                                                                                                                   | 7/20 [1:25:37<2:37:03, 724.85s/it]train - Global loss: 0.010388    Embedding norm: 1.0000   Triplets (all/active): 20.0/11.6
Pos dist (min/mean/max): 0.0405/0.0808/0.1471   Neg dist (min/mean/max): 0.7078/0.8492/0.9274
val - Global loss: 0.224147    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.8
Pos dist (min/mean/max): 0.0297/0.1163/0.2289   Neg dist (min/mean/max): 0.0372/0.1314/0.7385
 40%|███████████████████████████████████████████████████████████████████████▏                                                                                                          | 8/20 [1:37:34<2:24:27, 722.32s/it]train - Global loss: 0.005865    Embedding norm: 1.0000   Triplets (all/active): 20.0/5.9
Pos dist (min/mean/max): 0.0307/0.0635/0.1164   Neg dist (min/mean/max): 0.7115/0.8390/0.9331
val - Global loss: 0.160715    Embedding norm: 1.0000   Triplets (all/active): 20.0/399.3
Pos dist (min/mean/max): 0.0013/0.0485/0.1755   Neg dist (min/mean/max): 0.0012/0.0509/0.3290
 45%|████████████████████████████████████████████████████████████████████████████████                                                                                                  | 9/20 [1:49:30<2:12:02, 720.20s/it]train - Global loss: 0.004510    Embedding norm: 1.0000   Triplets (all/active): 20.0/6.1
Pos dist (min/mean/max): 0.0273/0.0480/0.0820   Neg dist (min/mean/max): 0.7157/0.7833/0.8847
val - Global loss: 0.135554    Embedding norm: 1.0000   Triplets (all/active): 20.0/397.8
Pos dist (min/mean/max): 0.0062/0.0252/0.0983   Neg dist (min/mean/max): 0.0060/0.0285/0.5567
 50%|████████████████████████████████████████████████████████████████████████████████████████▌                                                                                        | 10/20 [2:01:26<1:59:50, 719.08s/it]train - Global loss: 0.014267    Embedding norm: 1.0000   Triplets (all/active): 20.0/18.7
Pos dist (min/mean/max): 0.0429/0.0836/0.1479   Neg dist (min/mean/max): 0.6274/0.7919/0.8835
val - Global loss: 0.162442    Embedding norm: 1.0000   Triplets (all/active): 20.0/399.1
Pos dist (min/mean/max): 0.0018/0.0470/0.1755   Neg dist (min/mean/max): 0.0018/0.0517/0.8631
 55%|█████████████████████████████████████████████████████████████████████████████████████████████████▎                                                                               | 11/20 [2:13:22<1:47:42, 718.02s/it]train - Global loss: 0.010062    Embedding norm: 1.0000   Triplets (all/active): 20.0/11.8
Pos dist (min/mean/max): 0.0492/0.0908/0.1569   Neg dist (min/mean/max): 0.6495/0.8164/0.8854
val - Global loss: 0.180442    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.7
Pos dist (min/mean/max): 0.0025/0.0719/0.2188   Neg dist (min/mean/max): 0.0022/0.0789/1.0303
 60%|██████████████████████████████████████████████████████████████████████████████████████████████████████████▏                                                                      | 12/20 [2:25:18<1:35:39, 717.43s/it]train - Global loss: 0.004339    Embedding norm: 1.0000   Triplets (all/active): 20.0/6.9
Pos dist (min/mean/max): 0.0430/0.0814/0.1428   Neg dist (min/mean/max): 0.7358/0.8489/0.9293
val - Global loss: 0.224600    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.1
Pos dist (min/mean/max): 0.0045/0.1053/0.2455   Neg dist (min/mean/max): 0.0045/0.1083/0.3137
 65%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                                              | 13/20 [2:37:13<1:23:37, 716.74s/it]train - Global loss: 0.003673    Embedding norm: 1.0000   Triplets (all/active): 20.0/5.4
Pos dist (min/mean/max): 0.0385/0.0754/0.1338   Neg dist (min/mean/max): 0.7547/0.8723/0.9194
val - Global loss: 0.236573    Embedding norm: 1.0000   Triplets (all/active): 20.0/383.8
Pos dist (min/mean/max): 0.0056/0.1184/0.3399   Neg dist (min/mean/max): 0.0188/0.2360/1.2790
 70%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉                                                     | 14/20 [2:49:10<1:11:40, 716.72s/it]train - Global loss: 0.003818    Embedding norm: 1.0000   Triplets (all/active): 20.0/6.2
Pos dist (min/mean/max): 0.0392/0.0705/0.1197   Neg dist (min/mean/max): 0.7938/0.8608/0.9325
val - Global loss: 0.165365    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.3
Pos dist (min/mean/max): 0.0019/0.0518/0.2263   Neg dist (min/mean/max): 0.0019/0.0581/0.5769
 75%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                            | 15/20 [3:01:05<59:41, 716.34s/it]train - Global loss: 0.002332    Embedding norm: 1.0000   Triplets (all/active): 20.0/3.2
Pos dist (min/mean/max): 0.0321/0.0577/0.1026   Neg dist (min/mean/max): 0.7688/0.8284/0.9192
val - Global loss: 0.157235    Embedding norm: 1.0000   Triplets (all/active): 20.0/397.6
Pos dist (min/mean/max): 0.0016/0.0460/0.2090   Neg dist (min/mean/max): 0.0014/0.0562/0.8213
 80%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                                   | 16/20 [3:13:02<47:45, 716.43s/it]train - Global loss: 0.001819    Embedding norm: 1.0000   Triplets (all/active): 20.0/2.9
Pos dist (min/mean/max): 0.0299/0.0549/0.0977   Neg dist (min/mean/max): 0.7625/0.8511/0.9221
val - Global loss: 0.171820    Embedding norm: 1.0000   Triplets (all/active): 20.0/397.7
Pos dist (min/mean/max): 0.0018/0.0641/0.2281   Neg dist (min/mean/max): 0.0014/0.0801/0.9150
 85%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                          | 17/20 [3:24:57<35:48, 716.10s/it]train - Global loss: 0.001763    Embedding norm: 1.0000   Triplets (all/active): 20.0/1.2
Pos dist (min/mean/max): 0.0298/0.0547/0.0986   Neg dist (min/mean/max): 0.7842/0.8557/0.9254
val - Global loss: 0.181869    Embedding norm: 1.0000   Triplets (all/active): 20.0/397.9
Pos dist (min/mean/max): 0.0017/0.0724/0.2352   Neg dist (min/mean/max): 0.0015/0.0874/0.8872
 90%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                  | 18/20 [3:36:53<23:52, 716.18s/it]train - Global loss: 0.001372    Embedding norm: 1.0000   Triplets (all/active): 20.0/1.6
Pos dist (min/mean/max): 0.0290/0.0538/0.0960   Neg dist (min/mean/max): 0.7675/0.8377/0.9130
val - Global loss: 0.181953    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.8
Pos dist (min/mean/max): 0.0019/0.0728/0.2410   Neg dist (min/mean/max): 0.0014/0.0861/0.9684
 95%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████         | 19/20 [3:48:49<11:56, 716.03s/it]train - Global loss: 0.001379    Embedding norm: 1.0000   Triplets (all/active): 20.0/1.1
Pos dist (min/mean/max): 0.0287/0.0535/0.0960   Neg dist (min/mean/max): 0.7742/0.8435/0.9172
val - Global loss: 0.174906    Embedding norm: 1.0000   Triplets (all/active): 20.0/398.4
Pos dist (min/mean/max): 0.0017/0.0649/0.2312   Neg dist (min/mean/max): 0.0013/0.0786/0.9819
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [4:00:46<00:00, 722.32s/it]
Training completed.
We noticed that the validation loss (val loss) barely decreases. Is this normal? We are using the Docker environment as provided.

**Comment by hanjun815** [comment]
Hello,

Thank you for your interest in our research. We found a typo in the previous config/config.txt for the batch_size setting and have corrected it:

batch_size = 20 → batch_size = 256

Please try again with the updated configuration.

Thank you,
Hanjun Kim

**Comment by hanjun815** [comment]
Small batch sizes increase gradient noise, which can make training unstable and slow convergence. They also make BatchNorm’s per-batch statistics unreliable, potentially hurting accuracy. As workarounds, either (a) scale the learning rate to your batch size, or (b) use gradient accumulation so the global batch matches the paper’s setting.

Thank you,
Hanjun Kim

**Comment by hanjun815** [comment]
Thank you sincerely for your interest in my research. As you can see from the code, since we use random sampling for negative sampling in batch hard triplet loss, it’s natural that the inference results cannot be exactly the same.
```
neg_indices = np.where(pos_distances > self.neg_threshold)[0]
selected_neg_indices = np.random.choice(neg_indices, size=5, replace=True)
```
By adjusting _size_ and _neg_threshold_ here, you can likely achieve even better performance than what’s reported in the paper. (Even without optimizing parameters like size and neg_threshold, we confirmed significantly superior performance compared to other SOTA methods.)

And if you want even better performance, I recommend switching the backbone to ResNet34 — it dramatically boosts performance to the point of achieving 1.000 (Recall@1) on the Sports Complex 02 sequence. It’s already implemented in the code; just change the following line in SHeRLoc/models/radar/baseline_nets.py:

`self.fe = ResNetFeatureExtractor(resnet_type="resnet34", in_channels=1)`

Thanks,
Hanjun Kim
